hierarchicalforecast/core.py

```diff
@@ -90,170 +102,314 @@ class HierarchicalReconciliation:
         `num_samples`: int=-1, if positive return that many probabilistic coherent samples.
         `seed`: int=0, random seed for numpy generator's replicability.<br>
-        `sort_df` : bool (default=True), if True, sort `df` by [`unique_id`,`ds`].<br>
+        `sort_df` : deprecated.<br>
         `is_balanced`: bool=False, wether `Y_df` is balanced, set it to True to speed things up if `Y_df` is balanced.<br>
+        `id_col` : str='unique_id', column that identifies each serie.<br>
+        `time_col` : str='ds', column that identifies each timestep, its values can be timestamps or integers.<br>
+        `target_col` : str='y', column that contains the target.<br>
 
         **Returns:**<br>
-        `Y_tilde_df`: pd.DataFrame, with reconciled predictions.
+        `Y_tilde_df`: DataFrame, with reconciled predictions.
         """
+
         # Check input's validity and sort dataframes
-        Y_hat_df, S_df, Y_df, self.model_names = \
-                    self._prepare_fit(Y_hat_df=Y_hat_df,
-                                      S_df=S,
-                                      Y_df=Y_df,
-                                      tags=tags,
-                                      level=level,
-                                      intervals_method=intervals_method,
-                                      sort_df=sort_df)
+        is_pandas_Y_hat_df = nw.dependencies.is_pandas_like_dataframe(Y_hat_df)
+        is_pandas_S_df = nw.dependencies.is_pandas_like_dataframe(S)
+        if Y_df is not None:
+            is_pandas_Y_df = nw.dependencies.is_pandas_like_dataframe(Y_df)
+
+        Y_hat_df_nw, S_df_nw, Y_df_nw, self.model_names = self._prepare_fit(
+            Y_hat_df=Y_hat_df,
+            S_df=S,
+            Y_df=Y_df,
+            tags=tags,
+            level=level,
+            intervals_method=intervals_method,
+            id_col=id_col,
+            time_col=time_col,
+            target_col=target_col,
+        )
 
         # Initialize reconciler arguments
         reconciler_args = dict(
-            idx_bottom=S_df.index.get_indexer(S.columns),
-            tags={key: S_df.index.get_indexer(val) for key, val in tags.items()}
+            idx_bottom=np.arange(len(S_df_nw))[-S_df_nw.shape[1] :],
+            tags={
+                key: S_df_nw.with_columns(nw.col(id_col).is_in(val).alias("in_cols"))[
+                    "in_cols"
+                ]
+                .to_numpy()
+                .nonzero()[0]
+                for key, val in tags.items()
+            },
         )
 
         any_sparse = any([method.is_sparse_method for method in self.reconcilers])
+        S_df_nw_cols_ex_id_col = S_df_nw.columns
+        S_df_nw_cols_ex_id_col.remove(id_col)
         if any_sparse:
+            if not is_pandas_Y_hat_df or not is_pandas_S_df:
+                raise ValueError(
+                    "You have one or more sparse reconciliation methods. Please convert `S_df` and `Y_hat_df` to a pandas DataFrame."
```

Comment:
Does S have to be a dataframe? Can't we store it as a sparse matrix and just store the ids somewhere else? In my opinion we shouldn't even have dense methods, so not supporting sparse methods for polars seems inconvenient.

Reply:
At the moment, yes. The dense/sparse debate isn't something to solve in this PR. 

Polars doesn't support sparse.